File: solutions/elixir/food-chain/1/lib/food_chain.ex
defmodule FoodChain do
  @doc """
  Generate consecutive verses of the song 'I Know an Old Lady Who Swallowed a Fly'.
  """

  @animals [
    %{kind: "fly"},
    %{kind: "spider", verse: "It wriggled and jiggled and tickled inside her."},
    %{kind: "bird", verse: "How absurd to swallow a bird!"},
    %{kind: "cat", verse: "Imagine that, to swallow a cat!"},
    %{kind: "dog", verse: "What a hog, to swallow a dog!"},
    %{kind: "goat", verse: "Just opened her throat and swallowed a goat!"},
    %{kind: "cow", verse: "I don't know how she swallowed a cow!"},
    %{kind: "horse", verse: "She's dead, of course!"}
  ]

  @spec recite(start :: pos_integer, stop :: pos_integer) :: String.t()
  def recite(stop, stop) do
    (recite_verse(stop) |> Enum.join("\n")) <> "\n"
  end

  def recite(start, stop) do
    start..stop
    |> Enum.map(&recite(&1, &1))
    |> Enum.join("\n")
  end

  @spec recite_verse(index :: pos_integer) :: [String.t()]
  defp recite_verse(index) do
    animal = @animals |> Enum.at(index - 1)
    animal_verse = Map.get(animal, :verse)
    [start(animal.kind), animal_verse | verse(index - 1)] |> Enum.reject(&is_nil/1)
  end

  @spec verse(index :: non_neg_integer) :: [String.t()]
  defp verse(0) do
    ["I don't know why she swallowed the fly. Perhaps she'll die."]
  end

  defp verse(2) do
    [
      "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her."
      | verse(1)
    ]
  end

  defp verse(7) do
    []
  end

  defp verse(index) do
    current_animal = @animals |> Enum.at(index)
    previous_animal = @animals |> Enum.at(index - 1)
    [combine(current_animal.kind, previous_animal.kind) | verse(index - 1)]
  end

  defp start(animal_name) do
    "I know an old lady who swallowed a #{animal_name}."
  end

  defp combine(current_animal, previous_animal) do
    "She swallowed the #{current_animal} to catch the #{previous_animal}."
  end
end

File: solutions/python/food-chain/1/food_chain.py
ANIMALS = {
    1:('fly','I don\'t know why she swallowed the fly. Perhaps she\'ll die.', ''),
    2:('spider','It wriggled and jiggled and tickled inside her.', ' that wriggled and jiggled and tickled inside her'),
    3:('bird','How absurd to swallow a bird!', ''),
    4:('cat','Imagine that, to swallow a cat!', ''),
    5:('dog','What a hog, to swallow a dog!', ''),
    6:('goat','Just opened her throat and swallowed a goat!', ''),
    7:('cow','I don\'t know how she swallowed a cow!', ''),
    8:('horse','She\'s dead, of course!', '')
    
}

START = 'I know an old lady who swallowed a {}.'
COMBINED = 'She swallowed the {} to catch the {}{}.'

def get_verse(verse):
    verse_text = [START.format(ANIMALS[verse][0]), ANIMALS[verse][1]]

    if verse < 8:
        for i in reversed(range(2, verse + 1)):
            verse_text.append(COMBINED.format(ANIMALS[i][0], ANIMALS[i - 1][0], ANIMALS[i - 1][2]))
        if verse != 1:
            verse_text.append(ANIMALS[1][1])

    return verse_text

def recite(start, end):
    return_list = []

    for i in range(start, end + 1):
        if i != start:
            return_list.append('')
        return_list.extend(get_verse(i))

    return return_list